doc: Add LwM2M engine state machine

jarlamsa opened the pull request:
The state machine depicts the states of the lwm2m engine
and the spawned RD client events that will be forwarded
to lwm2m client if it has registered a callback-function.
Most of the events are just informs for the client and the
RD client underneath works and recovers automatically from
most cases.

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

Commit:
doc: Add LwM2M engine state machine

The state machine depicts the states of the lwm2m engine
and the spawned RD client events that will be forwarded
to lwm2m client if it has registered a callback-function.
Most of the events are just informs for the client and the
RD client underneath works and recovers automatically from
most cases.

Signed-off-by: Jarno Lamsa <jarno.lamsa@nordicsemi.no>

jarlamsa commented:
@peknis @SeppoTakalo please review.

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+LwM2M engine and application events
+***********************************
+
+The Zephyr LwM2M engine defines events that can be sent back to the application via callback
+functions. The engine state machine shows when the events are spawned. Events depicted in
+the diagram are listed in the table. The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.
```
```suggestion
The Zephyr LwM2M engine defines events that can be sent back to the application through callback
functions.
The engine state machine shows when the events are spawned.
Events depicted in the diagram are listed in the table.
The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+   * - 1
+     - BOOTSTRAP_REG_FAILURE
+     - Bootstrap registration failed.
+       Happens if timeout or failure in bootstrap registration.
```
```suggestion
       Occurs if there is a timeout or failure in bootstrap registration.
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+   * - 2
+     - BOOTSTRAP_REG_COMPLETE
+     - Bootstrap registration complete.
+       Happens after successful bootstrap registration.
```
```suggestion
       Occurs after successful bootstrap registration.
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+     - No actions needed
+   * - 3
+     - BOOTSTRAP_TRANSFER_COMPLETE
+     - Bootstrap finish command received from the server.
```
```suggestion
     - Bootstrap complete command received from the server.
```

SeppoTakalo replied:
The operation is really named `Bootstrap-Finish`
https://www.openmobilealliance.org/release/LightweightM2M/V1_1_1-20190617-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1_1-20190617-A.html#6-1-7-2-0-6172-Bootstrap-Finish-Operation

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+   * - 4
+     - REGISTRATION_FAILURE
+     - Registration to LwM2M server failed.
+       Happens if timeout or failure in the registration.
```
```suggestion
       Occurs if there is a timeout or failure in the registration.
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+   * - 5
+     - REGISTRATION_COMPLETE
+     - Registration to LwM2M server successful.
+       Happens after a successful registration reply from LwM2M server
```
```suggestion
       Occurs after a successful registration reply from the LwM2M server
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+   * - 6
+     - REG_UPDATE_FAILURE
+     - Registration update failed.
+       Happens if a timeout happens during registration update.
```
```suggestion
       Occurs if a there is a timeout during registration update.
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+   * - 7
+     - REG_UPDATE_COMPLETE
+     - Registration update completed.
+       Happens after successful registration update reply from LwM2M server.
```
```suggestion
       Occurs after successful registration update reply from the LwM2M server.
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+   * - 9
+     - DISCONNECT
+     - Disconnected from LwM2M server.
+       Happens if timeout during discussing with server.
```
```suggestion
       Occurs if there is a timeout during the dialogue with the server.
```

SeppoTakalo replied:
Maybe this should be `communication with the server`

peknis commented:
The image should be redrawn using Visio. Look for an example here: https://projecttools.nordicsemi.no/confluence/display/TECHDOC/State+diagram
And more figure style stuff here: https://projecttools.nordicsemi.no/confluence/display/TECHDOC/Figure+guide

SeppoTakalo commented:
> The image should be redrawn using Visio. Look for an example here...

This change is for public Zephyr project. I don't think internal Nordic guidelines apply. I would instead enforce the use of open source tools when submitting changes to open source community.

SeppoTakalo commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,93 @@ value of 1 is ok here).
+     - Disconnected from LwM2M server.
+       Happens if timeout during discussing with server.
+       Also triggered after deregistration has been done.
+     - No actions needed
```
Should there be a mention that if connection is required, application should restart the client.

jarlamsa commented:
Thank you for your reviews @SeppoTakalo @peknis, I think I have addressed your review comments and made the corrections in the document. Please re-review.

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,95 @@ value of 1 is ok here).
+LwM2M engine and application events
+***********************************
+
+The Zephyr LwM2M engine defines events that can be sent back to the application via callback
+functions.
```
```suggestion
The Zephyr LwM2M engine defines events that can be sent back to the application through callback
functions.
```

peknis commented on doc/connectivity/networking/api/lwm2m.rst:
```diff
@@ -410,6 +410,95 @@ value of 1 is ok here).
+functions.
+The engine state machine shows when the events are spawned.
+Events depicted in the diagram are listed in the table.
+The events are prefixed with `LWM2M_RD_CLIENT_EVENT_`.
```
```suggestion
The events are prefixed with ``LWM2M_RD_CLIENT_EVENT_``.
```

rlubos reviewed (approved):
LGTM, thanks!
